python/pyarrow/io.pxi

```diff
@@ -262,6 +271,158 @@ cdef class NativeFile:
+        number of bytes written
+        """
+
+        cdef:
+            int64_t bytes_read
+            uint8_t* buf
+
+        self._assert_readable()
+
+        cdef Buffer py_buf = py_buffer(b)
+        cdef int64_t buf_len = py_buf.size
+
+        buf = py_buf.buffer.get().mutable_data()
+
+        with nogil:
+            check_status(self.rd_file.get().Read(buf_len, &bytes_read, buf))
+
+        return bytes_read
+
+    def readline(self, size=None):
+        """Read and return a line of bytes from the file.
+
+        If size is specified, read at most size bytes.
+
+        Line terminator is always b"\\n".
+
+        TODO: Move this method to libarrow."""
+
+        cdef:
+            int64_t c_nbytes
+            int64_t bytes_to_read
+            int64_t bytes_read
+            int64_t linebreak_pos = 0
+            int64_t initial_pos
+            uint8_t* linebreak_ptr
+            PyObject* obj
+            shared_ptr[CBuffer] read_buffer
+            int64_t READ_BUFFER_SIZE = 4096
+
+        self._assert_readable()
+
+        if size is None:
+            c_nbytes = self.size() - self.tell()
+        else:
+            c_nbytes = size
+
+        # ReadAt moves position inconsistently, so save the initial position
+        check_status(self.rd_file.get().Tell(&initial_pos))
+
+        with nogil:
+            while linebreak_pos < c_nbytes:
+                if self.rd_file.get().supports_zero_copy():
+                    # we try to read the whole file, since this operation
+                    # is zero copy
+                    check_status(self.rd_file.get().GetSize(&bytes_to_read))
+                else:
+                    # we read only READ_BUFFER_SIZE next bytes which might
+                    # not be the ideal size for every file type
+                    bytes_to_read = READ_BUFFER_SIZE
+                check_status(
+                    self.rd_file.get().ReadB(bytes_to_read, &read_buffer))
+
+                bytes_read = read_buffer.get().size()
+
+                if bytes_read == 0:
+                    break
+
+                linebreak_ptr = <uint8_t*> \
+                    memchr(read_buffer.get().data(), b'\n', bytes_read)
+                if linebreak_ptr != NULL:
+                    # +1 because we want to read the newline char, too
+                    linebreak_pos += <int64_t> \
+                        (linebreak_ptr - read_buffer.get().data()) + 1
+                    break
+                else:
+                    linebreak_pos += bytes_read
+
+        if linebreak_pos > c_nbytes:
+            linebreak_pos = c_nbytes
+
+        obj = PyBytes_FromStringAndSizeNative(NULL, linebreak_pos)
+        cdef uint8_t* buf = <uint8_t*> cp.PyBytes_AS_STRING(<object> obj)
+
+        with nogil:
+            # ReadAt inconsistently moves the position, so do it manually
+            check_status(self.rd_file.get().Seek(initial_pos))
+
+            check_status(self.rd_file.get().Read(linebreak_pos,
```

Comment:
So we're reading again data that we've already read during the loop. Depending on the kind of file, this may be a system call or some other costly operation.

Reply:
Ideally we would ready piecewise into the bytes object and resize it at the end...